### ISSUE TITLE
Use Oracle PDBs as testcontainer

### ISSUE THREAD

**Issue by MeyerNils** [testcontainers/testcontainers-java#2369]
Docker images being currently provided by Oracle (https://github.com/oracle/docker-images/tree/master/OracleDatabase/SingleInstance) have quite long startup times (in my environment several minutes) as they're creating databases on startup. This makes them rather heavy to use in a testcontainers context. This can be avoided by building images with pre-build database as discussed as well in #2315 

As usage of official containers seems to be the preferred approach like discussed e.g. in #2357 that I'd prefer as well a completely different approach came to my mind. 

In its current versions Oracle supports the concept of a Container Database containing several isolated Pluggable Databases (PDB). Spinning up a pluggable database takes only few seconds and provides an isolated environment very similar to a "real" standalone database. 

I could imagine that making use of this concept could be beneficial in the context of testcontainers, too - every requests for a testcontainer spins up a PDB in a running Oracle Docker Container (if it's running) or first starts an Oracle Docker Container based on an official image if it's not running and then spins up a PDB. This PDB could be thrown away after the test again. The Oracle Docker Container could either by kept alive or only stopped (not removed) to allow much faster spin up of the testenvironment on the next run. 

Sadly XE is limited to three PDBs (and therefore limits the number of parallel runs in one Docker Container to three) but this would not be the case for the "big" licenses. And for quite some cases three parallel runs is not that bad. 

What do you think about this idea in general? It would require no special preparation of an environment (as official docker containers could be pulled on first start) but significantly improve preparation times for later runs. If you'd consider it interesting I could work out some more details and give a prototype a try that could potentially lead to a new module.

**Comment by kiview** [comment]
Hey @MeyerNils,
thanks for the idea. TBH I've never heard about PDBs, but might open up some very interesting features.

Right now I'd like us to first get some good support for the now released free to use XE binaries working before diving into other features.  Do you have a time measurement for the creation of a new PDB?

**Comment by MeyerNils** [comment]
Hey @kiview,
You‘re meaning the time needed for creating a new PDB in a running XE instance? I‘d say < 10 seconds, I‘m currently not near my development environment to provide you some precise numbers but could summarize them next week. 

Are you planning to build the “new” support based on the upcoming images mentioned in https://twitter.com/GeraldVenzl/status/1230673245193539585, perhaps even on a potentially coming pre-built database or as “own” image on quai.io as mentioned in #2357?

I think I once tried to build an image with a pre-build database, too, locally. I’ll double check next week and if so I can provide comparison numbers for starting and building a new database, starting a container with a running database and creating a PDB.

**Comment by MeyerNils** [comment]
Hey @kiview,
Some more information about my local tries. Potentially things are not comparable perfectly, as I was using the images I had available locally (e.g. SE and not XE) but I think the dimension is good enough. 
(Windows 10, i7-7600U, 32 GB RAM, SSD, Docker for Windows)

I tried
1. docker run oracle/database:19.3.0-se2 container without initialized database: **1124 seconds**
2. docker start container container from step 1 after stopping it (to get a feeling, how much time is spent in initializing the database): **14 seconds** --> database initialization is consuming most of the time (nothing new, just confirmed)
3. Create pluggable database within running container: **10 seconds**
4. docker run oracle/inizialized-database:12.2.0.1-NONCDB-NOVOLUME-se2 (a customized image I created in the past, having an already initialized database within the image): **22 seconds**

Stopping a container always took approx. 11 seconds, no matter which one, dropping the pluggable database took 5 seconds.

Main differences between the official oracle image (initializing the database on startup) and my custom image with the initialized database are following
- I did not expose $ORACLE_BASE/oradata as volume, as the database is initialized within this folder and I'd like to keep it in the image
- I extracted big parts from runOracle.sh, where initialization is triggered if needed, into an initDatabase.sh script, that will not block at the end and therefore could be used for image creation
As I would have needed some tries to get this working again as well with Oracle 19 I decided to stay with my existing Oracle 12.2 image for comparison to ease things for me.

Are those the time measures you were asking for?
